File: src/lib/tiposkripto/tests/calculator/Calculator.test.specification.ts
import type { ITestSpecification } from "../../src/CoreTypes";
import type { ICalculatorNode, O } from "./Calculator.test.types";

export const specification: ITestSpecification<ICalculatorNode, O> = (
  // Suite,
  Given,
  When,
  Then,
  Describe,
  It,
  Confirm,
  Value,
  Should,
) => {
  return [

    // TDT pattern: Confirm creates a BaseConfirm instance
    Confirm["addition"](
      [
        [Value.of([1, 1]), Should.beEqualTo(2222)],
        [Value.of([2, 3]), Should.beGreaterThan(4)],
      ],
      ["./Readme.md3"],
    ),

    // AAA pattern: Describe creates a BaseDescribe instance
    Describe["another simple calculator"]("some input")(
      [
        It["can save 1 memory"](),
        It["can save 2 memories"](),
      ],
      ["./Readme.md"],
    ),

    // BDD pattern: Given creates a BaseGiven instance
    Given.Default("some input")(
      [
        When.press("5"),
        When.press("+"),
        When.press("3"),
        When.enter(),
      ],
      [Then.result("8")],
      ["./Readme.md"],
    ),


    Confirm["addition"](
      [
        [Value.of([3, 3]), Should.beEqualTo(3)],
      ],
      ["./Readme.md"],
    ),

    Confirm["addition"](
      [
        [Value.of([3, 32]), Should.beEqualTo(32)],
      ],
      ["./Readme.md"],
    ),

    Confirm["addition"](
      [
        [Value.of([3, 332]), Should.beEqualTo(332)],
      ],
      ["./Readme.md"],
    ),

  ];
}

File: src/lib/rubeno/examples/calculator/Calculator.test.specification.rb
module Rubeno
  module Examples
    module CalculatorTestSpecification
      def self.call(given_wrapper, when_wrapper, then_wrapper, 
                    describes_wrapper, its_wrapper, 
                    confirms_wrapper, values_wrapper, shoulds_wrapper)
        [
          # TDT pattern: Confirm creates a BaseConfirm instance
          confirms_wrapper.addition.().(
            [
              [values_wrapper.of.([1, 1]), shoulds_wrapper.be_equal_to.(2222)],
              [values_wrapper.of.([2, 3]), shoulds_wrapper.be_greater_than.(4)],
            ],
            ["./Readme.md"],
          ),
          
          # AAA pattern: Describe creates a BaseDescribe instance
          describes_wrapper.a_simple_calculator.("some input").(
            [
              its_wrapper.can_save_1_memory.(),
              its_wrapper.can_save_2_memories.(),
            ],
            ["./Readme.md"],
          ),
          
          # BDD pattern: Given creates a BaseGiven instance
          given_wrapper.default.("some input").(
            [
              when_wrapper.press.("5"),
              when_wrapper.press.("+"),
              when_wrapper.press.("3"),
              when_wrapper.enter.(),
            ],
            [then_wrapper.result.("8")],
            ["./Readme.md"],
          ),
          
          confirms_wrapper.addition.().(
            [
              [values_wrapper.of.([3, 3]), shoulds_wrapper.be_equal_to.(3)],
            ],
            ["./Readme.md"],
          ),
          
          confirms_wrapper.addition.().(
            [
              [values_wrapper.of.([3, 32]), shoulds_wrapper.be_equal_to.(32)],
            ],
            ["./Readme.md"],
          ),
          
          confirms_wrapper.addition.().(
            [
              [values_wrapper.of.([3, 332]), shoulds_wrapper.be_equal_to.(332)],
            ],
            ["./Readme.md"],
          ),
        ]
      end
    end
  end
end
